components/director/internal/domain/formationconstraint/operators/asynchronous_flow_control_operator.go

```diff
@@ -0,0 +1,124 @@
+func AsynchronousFlowControlOperatorInput() OperatorInput {
+	return &formationconstraint.AsynchronousFlowControlOperatorInput{}
+}
+
+// AsynchronousFlowControlOperator is an operator that based on different conditions behaves like the redirect operator, it redirects the formation assignment notification.
+// In other cases it mutates the state, in order to control the flow of the engine, so that the assignment doesn't get deleted too early,
+// and it resends the notification to the redirection endpoint, so that it can finish the cleanup.
+// It introduces new deleting states.
+func (e *ConstraintEngine) AsynchronousFlowControlOperator(ctx context.Context, input OperatorInput) (bool, error) {
+	log.C(ctx).Infof("Starting executing operator: %s", AsynchronousFlowControlOperator)
+
+	defer func() {
+		if err := recover(); err != nil {
+			log.C(ctx).WithField(logrus.ErrorKey, err).Panic("recovered panic")
+			debug.PrintStack()
+		}
+	}()
+
+	ri, ok := input.(*formationconstraint.AsynchronousFlowControlOperatorInput)
+	if !ok {
+		return false, errors.Errorf("Incompatible input for operator: %s", AsynchronousFlowControlOperator)
+	}
+
+	log.C(ctx).Infof("Enforcing constraint on resource of type: %q and subtype: %q for location with constraint type: %q and operation name: %q during %q operation", ri.ResourceType, ri.ResourceSubtype, ri.Location.ConstraintType, ri.Location.OperationName, ri.Operation)
+
+	if ri.Operation == model.AssignFormation && ri.Location.OperationName == model.SendNotificationOperation && ri.Location.ConstraintType == model.PreOperation {
+		return e.RedirectNotification(ctx, &ri.RedirectNotificationInput)
+	}
+	if ri.Operation == model.UnassignFormation && ri.Location.OperationName == model.SendNotificationOperation && ri.Location.ConstraintType == model.PreOperation {
+		formationAssignment, err := RetrieveFormationAssignmentPointer(ctx, ri.FAMemoryAddress)
+		if err != nil {
+			return false, err
+		}
+		if formationAssignment.State == string(model.InstanceCreatorDeletingAssignmentState) ||
+			formationAssignment.State == string(model.InstanceCreatorDeleteErrorAssignmentState) {
+			log.C(ctx).Infof("Tenant mapping participant processing unassign notification has alredy finished, redirecting notification for assignment %q with state %q to instance creator", formationAssignment.ID, formationAssignment.State)
+			ri.ShouldRedirect = true
+			return e.RedirectNotification(ctx, &ri.RedirectNotificationInput)
+		}
+		return true, nil
+	}
+
+	if ri.Location.OperationName == model.NotificationStatusReturned && ri.Location.ConstraintType == model.PreOperation {
+		formationAssignment, err := RetrieveFormationAssignmentPointer(ctx, ri.FAMemoryAddress)
+		if err != nil {
+			return false, err
+		}
+		reverseAssignment, err := RetrieveFormationAssignmentPointer(ctx, ri.ReverseFAMemoryAddress)
+		if err != nil {
+			log.C(ctx).Warnf(errors.Wrapf(err, "Reverse assignment not found").Error())
```

Comment:
Is there a use case for this error that could potentially break some stuff if the flow is not stopped with a return statement?

Reply:
The reverse assignment could be `nil` in the one case that it is used, that's why it is not stopped with a return statement. 
The `GenerateFormationAssignmentPair` method handles `nil` reverse assignment, because the reverse could have been deleted at the time of generating notifications for unassign. Looking at the code now this can be moved in the one if that is used.